sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java

```diff
@@ -233,21 +233,23 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
 
     @Override
     public PCollection<KinesisRecord> expand(PBegin input) {
-      org.apache.beam.sdk.io.Read.Unbounded<KinesisRecord> read =
+      Unbounded<KinesisRecord> unbounded =
           org.apache.beam.sdk.io.Read.from(
-              new KinesisSource(getAWSClientsProvider(), getStreamName(),
-                  getInitialPosition(), getUpToDateThreshold()));
-      if (getMaxNumRecords() > 0) {
-        BoundedReadFromUnboundedSource<KinesisRecord> bounded =
-            read.withMaxNumRecords(getMaxNumRecords());
-        return getMaxReadTime() == null
-            ? input.apply(bounded)
-            : input.apply(bounded.withMaxReadTime(getMaxReadTime()));
-      } else {
-        return getMaxReadTime() == null
-            ? input.apply(read)
-            : input.apply(read.withMaxReadTime(getMaxReadTime()));
+              new KinesisSource(
+                  getAWSClientsProvider(),
+                  getStreamName(),
+                  getInitialPosition(),
+                  getUpToDateThreshold()));
+
+      PTransform<PBegin, PCollection<KinesisRecord>> transform = unbounded;
+
+      if (getMaxNumRecords() < Long.MAX_VALUE) {
+        transform = unbounded.withMaxNumRecords(getMaxNumRecords());
```

Comment:
The previous code supported a combination of both options, but the current code does not. Is this intentional?

Reply:
I changed it to make it consistent with all the other uses on unbounded sources e.g. KafkaIO, MqttIO, etc.

Notice that the with methods on `Read.Unbounded` seem to be exclusive. See `org.apache.beam.sdk.io.Read.Unbounded#withMaxNumRecords` (and the method below), but they can be of course overwritten because `BoundedReadFromUnboundedSource` supports both parameters.

I suppose that a user that wants the combined options can just use use the time based one followed by `Sample.any`. What do you prefer? Should I let it like it was or make it consistent with the other IOs?

Reply:
Seems reasonable. Then can you just add code to verify that the user didn't specify both options at the same time, i.e. that we're not silently ignoring one of them?

Reply:
done.